[SPARK-26089][CORE] Handle corruption in large shuffle blocks

ankuriitg opened the pull request:
## What changes were proposed in this pull request?

SPARK-4105 added corruption detection in shuffle blocks but that was limited to blocks which are
smaller than maxBytesInFlight/3. This commit adds upon that by adding corruption check for large
blocks. There are two changes/improvements that are made in this commit:

1. Large blocks are checked upto maxBytesInFlight/3 size in a similar way as smaller blocks, so if a
large block is corrupt in the starting, that block will be re-fetched and if that also fails,
FetchFailureException will be thrown.
2. If large blocks are corrupt after size maxBytesInFlight/3, then any IOException thrown while
reading the stream will be converted to FetchFailureException.  This is slightly more aggressive
than was originally intended but since the consumer of the stream may have already read some records and processed them, we can't just re-fetch the block, we need to fail the whole task. Additionally, we also thought about maybe adding a new type of TaskEndReason, which would re-try the task couple of times before failing the previous stage, but given the complexity involved in that solution we decided to not proceed in that direction.

Thanks to @squito for direction and support.

## How was this patch tested?

Changed the junit test for big blocks to check for corruption.

ankuriitg committed:
[SPARK-26089][CORE] Handle corruption in large shuffle blocks

SPARK-4105 added corruption detection in shuffle blocks but that was limited to blocks which are
smaller than maxBytesInFlight/3. This commit adds upon that by adding corruption check for large
blocks. There are two changes/improvements that are made in this commit:

1. Large blocks are checked upto maxBytesInFlight/3 size in a similar way as smaller blocks, so if a
large block is corrupt in the starting, that block will be re-fetched and if that also fails,
FetchFailureException will be thrown.
2. If large blocks are corrupt after size maxBytesInFlight/3, then any IOException thrown while
reading the stream will be converted to FetchFailureException.  This is slightly more aggressive
than was originally intended but since the consumer of the stream may have already read some records
and processed them, we can't just re-fetch the block, we need to fail the whole task. Additionally,
we also thought about maybe adding a new type of TaskEndReason, which would re-try the task couple
of times before failing the previous stage, but given the complexity involved in that solution we
decided to not proceed in that direction.

Thanks to @squito for direction and support.

Testing Done:
Changed the junit test for big blocks to check for corruption.

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #100746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100746/testReport)** for PR 23453 at commit [`08da21f`](https://github.com/apache/spark/commit/08da21f4afda762280c2be11fc51cc07b66a964d).

SparkQA commented:
**[Test build #100746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100746/testReport)** for PR 23453 at commit [`08da21f`](https://github.com/apache/spark/commit/08da21f4afda762280c2be11fc51cc07b66a964d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100746/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?

ankuriitg committed:
Correct indentation

SparkQA commented:
**[Test build #100750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100750/testReport)** for PR 23453 at commit [`ca7e804`](https://github.com/apache/spark/commit/ca7e804322056ea658c3202894337165cde822f8).

squito commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
+      if (closeStreams) {
+        try {
+          if (count < maxSize) {
+            in.close()
```
maxSize might be exactly the same length as the stream, in which case you'll read to the end, but not close it.

ankuriitg replied:
Right, I left that edge case, as it will be handled by the concatenated stream later

squito commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
+   * Copy all data from an InputStream to an OutputStream upto maxSize and
+   * closes the input stream if closeStreams is true and all data is read.
+   */
+  def copyStreamUpto(
```
super nit: capital T, UpTo

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -453,14 +454,21 @@ final class ShuffleBlockFetcherIterator(
-            if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {
+            if (detectCorrupt && !input.eq(in)) {
               isStreamCopied = true
+              streamCompressedOrEncrypted = true
```
I wonder if we should do the inputStream checks even if `detectCorrupt` is false, and perhaps even change that default.  Or add another flag just for doing this memory copy, as its more expensive.

Also I think the comments around here need to be updated to explain what is happening now.

ankuriitg replied:
Sure, I can change it to always detect corruption. Which default did you mean though, the default value of `detectCorrupt` is true.

squito reviewed (commented):
just a partial review so far

there should be a test that a large block is read correctly, with the partial buffering & stream concatenation.  I think most unit tests will not end up creating a large block.

SparkQA commented:
**[Test build #100750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100750/testReport)** for PR 23453 at commit [`ca7e804`](https://github.com/apache/spark/commit/ca7e804322056ea658c3202894337165cde822f8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100750/
Test PASSed.

ankuriitg committed:
Review Comments: Part 1

1. Updated comments in the code
2. If IOException is thrown while reading from a stream, it will always be converted to a FetchFailureException, even when detectCorruption is false
3. Added a junit test which verifies that data can be read from concatenated stream

SparkQA commented:
**[Test build #100895 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100895/testReport)** for PR 23453 at commit [`e638c4e`](https://github.com/apache/spark/commit/e638c4edd24c4f9d9fee28e4ff684b8a91cd866e).

SparkQA commented:
**[Test build #100895 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100895/testReport)** for PR 23453 at commit [`e638c4e`](https://github.com/apache/spark/commit/e638c4edd24c4f9d9fee28e4ff684b8a91cd866e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100895/
Test FAILed.

ankuriitg committed:
Replaced 'getArgumentAt()' with 'getArguments()' because of changes in Mockito api across versions

SparkQA commented:
**[Test build #100898 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100898/testReport)** for PR 23453 at commit [`0d10626`](https://github.com/apache/spark/commit/0d1062631ca57c3a2aa2280f2a4b8756818031fa).

SparkQA commented:
**[Test build #100898 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100898/testReport)** for PR 23453 at commit [`0d10626`](https://github.com/apache/spark/commit/0d1062631ca57c3a2aa2280f2a4b8756818031fa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100898/
Test PASSed.

rezasafi commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              isStreamCopied = Utils.copyStreamUpTo(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
```
What is the reasoning behind considering one-third of the block, why for example not one-fourth?

ankuriitg replied:
That limit was already defined. I did not change that limit. The purpose was to limit the amount of memory used but I am not sure how we came up with that exact threshold.

rezasafi replied:
I can't find where it is defined. I thought your change is proposing that. Sorry if I missed that

ankuriitg replied:
Look for this line on left side of the files changed: `if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {`

rezasafi commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
+              if (isStreamCopied) {
+                input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              } else {
```
What is the behavior here if the stream wasn't copied because of the corruption? It seems that this else ignores the corruption in one-third and will try to see if the corruption is happening  in the remaining two-third which seems not necessary.

ankuriitg replied:
I am guessing you did not look into Utils.copyStreamUpTo(). isStreamCopied is true when the size of the stream is less than maxBytesInFlight/3 and that stream was copied. If it is longer, then it will go to the else block. If stream wasn't copied because of corruption, it will be handled in catch block.

rezasafi replied:
A little confusing if this if only check the size but its condition only says isStreamCopied.

ankuriitg replied:
Please look at `Utils.copyStreamUpTo()`

rezasafi commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -490,7 +499,9 @@ final class ShuffleBlockFetcherIterator(
 
     currentResult = result.asInstanceOf[SuccessFetchResult]
-    (currentResult.blockId, new BufferReleasingInputStream(input, this))
+    (currentResult.blockId,
```
What is the purpose of this new updated return? currentResult.blockId is already returning in tuple.

ankuriitg replied:
It is still returning the same tuple. I am just passing more arguments to the constructor of BufferReleasingInputStream

rezasafi reviewed (commented):
Thanks for the work here. This is just an initial review to better understand your approach. I think It will be great to ping the original authors of SPARK-4105 and related changes to get their feedback as well.

ankuriitg commented:
cc @davies since you worked on SPARK-4105

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
+              } else {
+                input = new SequenceInputStream(
+                  out.toChunkedByteBuffer.toInputStream(dispose = true), input)
+              }
```
I think it would be better if you refactored this so `copyStreamUpTo` includes this part as well, and just returns an InputStream.  The main reason is for testing -- I worry about edge cases here.   If this is factored out, it would be much easier to test that the resulting input stream is OK if eg. there is an empty input, or just up to the limit, or over the limit by X, and make sure streams get closed correctly etc.

squito commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
+      closeStreams: Boolean = false): Boolean = {
+    var count = 0L
+    tryWithSafeFinally {
+      val bufSize = 8192
```
max(8192, maxSize)

squito reviewed (commented):
>> I wonder if we should do the inputStream checks even if detectCorrupt is false, and perhaps even change that default. Or add another flag just for doing this memory copy, as its more expensive.

> Which default did you mean though, the default value of detectCorrupt is true.

Sorry I didn't explain that very well.  What I was wondering is we should change the behavior to

* max the copy-into-memory check turned off by default
* but *always* do the on-the-fly check, even if detectCorrupt is false.  You could add another option for this also, and have it default true, if really want to allow users to turn off all detection.

I'm saying this because I think that I'd want to run my spark apps with the new default I'm suggesting -- don't do the copy into memory, but do the on-the-fly check all the time, of the entire block.

ankuriitg committed:
Review comments: Part 2

1. Minor changes
2. Added a new config for detecting corruption by using extra memory with
default set to false
3. Added test cases for copyStreamUpTo

SparkQA commented:
**[Test build #101777 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101777/testReport)** for PR 23453 at commit [`abbec63`](https://github.com/apache/spark/commit/abbec63212d2ffc073b8fca6e391b1e2ca2b8dd0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101777/
Test FAILed.

SparkQA commented:
**[Test build #101776 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101776/testReport)** for PR 23453 at commit [`d40d396`](https://github.com/apache/spark/commit/d40d396a2c83d5c92f685114a332da0058ade92a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101776/
Test FAILed.

ankuriitg commented:
retest this please

SparkQA commented:
**[Test build #101832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101832/testReport)** for PR 23453 at commit [`abbec63`](https://github.com/apache/spark/commit/abbec63212d2ffc073b8fca6e391b1e2ca2b8dd0).

SparkQA commented:
**[Test build #101832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101832/testReport)** for PR 23453 at commit [`abbec63`](https://github.com/apache/spark/commit/abbec63212d2ffc073b8fca6e391b1e2ca2b8dd0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101832/
Test FAILed.

squito commented:
real build error:

```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/internal/config/package.scala:931: SHUFFLE_DETECT_CORRUPT is already defined as value SHUFFLE_DETECT_CORRUPT
[error]   private[spark] val SHUFFLE_DETECT_CORRUPT =
[error]                      ^
[error] one error found
```

ankuriitg committed:
Fixed compilation errors after merging with master

SparkQA commented:
**[Test build #101837 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101837/testReport)** for PR 23453 at commit [`847898d`](https://github.com/apache/spark/commit/847898dd7cf826cda822049fe3ec8eb632b19040).

ankuriitg commented:
> real build error:
> 
> ```
> [error] /home/jenkins/workspace/SparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/internal/config/package.scala:931: SHUFFLE_DETECT_CORRUPT is already defined as value SHUFFLE_DETECT_CORRUPT
> [error]   private[spark] val SHUFFLE_DETECT_CORRUPT =
> [error]                      ^
> [error] one error found
> ```

Thanks, I realized it after this morning's build failure. The previous test failure confused me and I did not see the build failure before that.

SparkQA commented:
**[Test build #101837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101837/testReport)** for PR 23453 at commit [`847898d`](https://github.com/apache/spark/commit/847898dd7cf826cda822049fe3ec8eb632b19040).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101837/
Test PASSed.

squito reviewed (commented):
just some test updates.

@tgravescs how do you feel about my suggested updates to the defaults?

tgravescs commented:
I think you are talking about:

>  but always do the on-the-fly check, even if detectCorrupt is false. You could add another option for this also, and have it default true, if really want to allow users to turn off all detection.

so the only place detectCorrupt is used after this change is  passed into BufferReleasingInputStream 
 for the read, and its default is true.  If we want to make the exception handling there always present, are you saying just remove the detectCorrupt config or are you suggesting something with the streamCompressedOrEncrypted?

squito commented:
sorry my earlier comment wasn't clear -- ankur was confused too but I think this version of the patch implements what I think the right option is:

* by default, do not copy into memory at all, but do on-the-fly detection of corruption.  This alway results in a fetch-failure
* optionally, go back to the old behavior, to copy into memory if the block is small.  Then you'd first retry pulling the data, and throw a fetch failure after that.  Turned on with "spark.shuffle.detectCorrupt.useExtraMemory=true"
* optionally, never do any check, even on the fly (using the old config, "spark.shuffle.detectCorrupt=false")

tgravescs commented:
Ok, That makes sense to me and agree with you.

ankuriitg committed:
Review comments: Part 3

1. Changed test to also compare the contents of stream
2. Other minor refactoring

SparkQA commented:
**[Test build #101996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101996/testReport)** for PR 23453 at commit [`17e36c8`](https://github.com/apache/spark/commit/17e36c89ad954dd66149dce45424a9657cee5a11).

SparkQA commented:
**[Test build #101997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101997/testReport)** for PR 23453 at commit [`c8d3569`](https://github.com/apache/spark/commit/c8d356909f86bedc1015246cd23bc70ebe308e16).

squito commented on core/src/test/scala/org/apache/spark/util/UtilsSuite.scala:
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
+
+    val limit = 1000
+    // testing for inputLength less than, equal to and greater than limit
+    List(900, 1000, 1100).foreach { inputLength =>
```
sorry I'm being very nitpicky here, but I really want to make sure we're testing the exact edge conditions when the limit is -2, -1, 0, +1, +2 of the full length.  Its OK if you don't free the buffer till you are past it, but it should happen shortly after the limit.

squito commented on core/src/test/scala/org/apache/spark/util/UtilsSuite.scala:
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
+  }
+
+  private def getFieldValue(obj: AnyRef, fieldName: String): Any = {
+    val mirror = universe.runtimeMirror(obj.getClass().getClassLoader())
```
Use java reflection instead of scala reflection here -- its known to more developers, has fewer strange gotchas, simpler and is sufficient for this.

squito commented on core/src/test/scala/org/apache/spark/util/UtilsSuite.scala:
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
+      val in = new ByteArrayInputStream(bytes.take(inputLength))
+      val (fullyCopied: Boolean, mergedStream: InputStream) = Utils.copyStreamUpTo(in, limit, true)
+      try {
+        val byteBufferInputStream = if (mergedStream.isInstanceOf[ChunkedByteBufferInputStream]) {
```
Add some comments here, like "Get a handle on the buffered data, to make sure the memory gets freed once we read past the end of it.  Need to use refelction to get handle on inner structures for this check"

squito commented on core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:
```diff
@@ -461,10 +485,68 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
+    // Only one block should be returned which has corruption after maxBytesInFlight/3
+    val (id, st) = iterator.next()
+    assert(id === ShuffleBlockId(0, 1, 0))
+    intercept[FetchFailedException] { iterator.next(); iterator.next() }
```
I'm a little confused here -- why do you need to call iterator.next() twice?

ankuriitg committed:
Review comments: Part 4

SparkQA commented:
**[Test build #101999 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101999/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).

SparkQA commented:
**[Test build #101996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101996/testReport)** for PR 23453 at commit [`17e36c8`](https://github.com/apache/spark/commit/17e36c89ad954dd66149dce45424a9657cee5a11).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101996/
Test FAILed.

SparkQA commented:
**[Test build #101999 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101999/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101999/
Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7684/
Test PASSed.

SparkQA commented:
**[Test build #102008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102008/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).

SparkQA commented:
**[Test build #102008 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102008/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102008/
Test FAILed.

SparkQA commented:
**[Test build #102037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102037/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).

SparkQA commented:
**[Test build #102037 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102037/testReport)** for PR 23453 at commit [`80d4e6b`](https://github.com/apache/spark/commit/80d4e6bf1d77cb0785d3ff37bca10552691d4e47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102037/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -335,6 +336,47 @@ private[spark] object Utils extends Logging {
+  def copyStreamUpTo(
+      in: InputStream,
+      maxSize: Long,
+      closeStreams: Boolean = false): (Boolean, InputStream) = {
```
I don't think you need to expose `closeStreams`, just have it always be true.  At least I can't think of a reason why it should default to `false` (`copyStream` is used more places ...)

ankuriitg committed:
Review comments: Part 5

Changes to unit test case

SparkQA commented:
**[Test build #102065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102065/testReport)** for PR 23453 at commit [`c0c550f`](https://github.com/apache/spark/commit/c0c550f13784b1ae193ab0e9be3d42fcea36a34d).

SparkQA commented:
**[Test build #102065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102065/testReport)** for PR 23453 at commit [`c0c550f`](https://github.com/apache/spark/commit/c0c550f13784b1ae193ab0e9be3d42fcea36a34d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102065/
Test PASSed.

ankuriitg committed:
Review comments: Part 6

SparkQA commented:
**[Test build #102223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102223/testReport)** for PR 23453 at commit [`016d0d7`](https://github.com/apache/spark/commit/016d0d71fa173a74a9bc5d7a4c822988fe8fd444).

SparkQA commented:
**[Test build #102223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102223/testReport)** for PR 23453 at commit [`016d0d7`](https://github.com/apache/spark/commit/016d0d71fa173a74a9bc5d7a4c822988fe8fd444).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102223/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -571,7 +582,8 @@ final class ShuffleBlockFetcherIterator(
 
-  private def throwFetchFailedException(blockId: BlockId, address: BlockManagerId, e: Throwable) = {
+  private[storage] def throwFetchFailedException(
+      blockId: BlockId, address: BlockManagerId, e: Throwable) = {
```
nit: if the method declaration is multi-line, each arg on its own line.  And this method should have had a return type in the first place (probably my fault, oops)

```scala
private[storage] def throwFetchFailedException(
    blockId: BlockId,
    address: BlockManagerId,
    e: Throwable): Unit = {
```

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -583,15 +595,28 @@ final class ShuffleBlockFetcherIterator(
+    private val iterator: ShuffleBlockFetcherIterator,
+    private val blockId: BlockId,
+    private val address: BlockManagerId,
+    private val streamCompressedOrEncrypted: Boolean)
```
I would rename `streamCompressedOrEncrypted` to `detectCorruption` or something like that, as the condition is a bit more complex now (you also check the `detectCorrupt` config when passing this in)

squito commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -335,6 +336,44 @@ private[spark] object Utils extends Logging {
+  def copyStreamUpTo(in: InputStream, maxSize: Long): (Boolean, InputStream) = {
+    var count = 0L
+    val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
+    val streamCopied = tryWithSafeFinally {
```
nit: I think a better name for `streamCopied` would be `fullyCopied` ... everytime I come back to this code I always get a bit confused that maybe no copying is happening in some situations.

squito commented on core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:
```diff
@@ -425,28 +450,97 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
+
+    // Only one block should be returned which has corruption after maxBytesInFlight/3 because the
+    // other block will detect corruption on first fetch, and then get added to the queue again for
+    // a retry
```
I'd reword this -- you get one block because you call next() once.  You really want to explain why you get a certain block back.

```We'll get back the block which has corruption after maxBytesInFlight/3 because ...```

(assuming I understand this correctly...)

squito commented on core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:
```diff
@@ -425,28 +450,97 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
+    }
+
+    // Following will succeed as it reads part of the stream which is not corrupt. This will read
+    // maxBytesInFlight/3 bytes from first stream and remaining from the second stream
```
... will read maxBytesInFlight/3 bytes from the portion copied into memory, and the remaining from the underlying stream.

squito commented on core/src/test/scala/org/apache/spark/util/UtilsSuite.scala:
```diff
@@ -211,6 +213,55 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
+
+    val limit = 1000
+    // testing for inputLength less than, equal to and greater than limit
+    List(998, 999, 1000, 1001, 1002).foreach { inputLength =>
```
`(limit - 2 to limit + 2).foreach`

squito reviewed (approved):
a few very minor style things, otherwise lgtm

ankuriitg committed:
Review Comments: Part 7

Minor changes to variable names and comments

SparkQA commented:
**[Test build #103217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103217/testReport)** for PR 23453 at commit [`b178f65`](https://github.com/apache/spark/commit/b178f657af054511f3cf3ea06c953828dbce4831).

SparkQA commented:
**[Test build #103217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103217/testReport)** for PR 23453 at commit [`b178f65`](https://github.com/apache/spark/commit/b178f657af054511f3cf3ea06c953828dbce4831).
 * This patch **fails Java style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103217/
Test FAILed.

SparkQA commented:
**[Test build #103218 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103218/testReport)** for PR 23453 at commit [`bd1a813`](https://github.com/apache/spark/commit/bd1a813216f1843f8c4d4c52749a27fccec0da5f).

SparkQA commented:
**[Test build #103218 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103218/testReport)** for PR 23453 at commit [`bd1a813`](https://github.com/apache/spark/commit/bd1a813216f1843f8c4d4c52749a27fccec0da5f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103218/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
-              // Decompress the whole block at once to detect any corruption, which could increase
-              // the memory usage tne potential increase the chance of OOM.
+              // Decompress the block upto maxBytesInFlight/3 at once to detect any corruption which
+              // could increase the memory usage and potentially increase the chance of OOM.
```
this comment & the one just above the `if` are redundant and a little bit wrong -- I think you only need one comment (don't care whether its above or below the if) and should be something more like:

We optionally decompress the first maxBytesInFlight/3 bytes into memory, to check for corruption in that portion of the data.  But even if that configuration is off, or if the corruption is later, we'll still try to detect the corruption later in the stream.

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              val (fullyCopied: Boolean, mergedStream: InputStream) = Utils.copyStreamUpTo(
+                input, maxBytesInFlight / 3)
```
I'm trying to understand why the

```scala
finally {
            // TODO: release the buf here to free memory earlier
            if (isStreamCopied) {
              in.close()
            }
```

is needed down below.  To be honest, I don't think it was needed in the old code.  The old `Utils.copyStream` was always called with `closeStreams=true`, and that would always close the input in a `finally` itself:

https://github.com/apache/spark/blob/d9978fb4e4d4de3a320b012373c18bd278462780/core/src/main/scala/org/apache/spark/util/Utils.scala#L302-L307

https://github.com/apache/spark/blob/d9978fb4e4d4de3a320b012373c18bd278462780/core/src/main/scala/org/apache/spark/util/Utils.scala#L330-L332

It doesn't hurt, but it also makes things unnecessarily confusing.  If you didn't need to do that `in.close()` below, you woudln't need to track `isStreamCopied`, and wouldn't need to even return `fullyCopied` from `Utils.copyStreamUpTo`.  That's really the part of this which is bugging me -- something seems off that we need to know whether or not the stream is fully copied, seems like it shouldn't matter.  If it does matter, aren't we getting something wrong in the case where the stream is *exactly* maxBytesInFlight / 3, but we haven't realized its fully copied because we haven't read past the end yet?

ankuriitg replied:
I think in.close() is needed if there is an exception while creating a wrapped stream. The first time I saw `isStreamCopied`, I was also confused and by looking at it more closely now, I realize that it is not doing what it is supposed to do.

I have removed `isStreamCopied` and instead used another condition to close the stream. Please check and let me know if it makes sense.

squito commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
+              val (fullyCopied: Boolean, mergedStream: InputStream) = Utils.copyStreamUpTo(
+                input, maxBytesInFlight / 3)
+              isStreamCopied = fullyCopied
+              input = mergedStream
```
not related to your changes, but while you're touching this file, can you add 2 more spaces of indentation to `|| corruptedBlocks.contains(blockId)) {` a couple of lines below this?

squito reviewed (commented):
Thanks for the update @ankuriitg .  sorry I am being very particular about this, its just a really core piece.  I have a couple of minor updates to comments -- but really I was doing one final pass and I got pretty confused about the existing `in.close()`.  I'd like us to clean that up now, while we're looking at this, if possible.   Left a longer comment inline, please check my reasoning.

ankuriitg committed:
Review Comments: Part 8

1. Ensured input stream is closed on exception
2. Minor comments changes

SparkQA commented:
**[Test build #103338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103338/testReport)** for PR 23453 at commit [`d36c862`](https://github.com/apache/spark/commit/d36c86251946e88a36e2e977d8ae1a0ae815a8d5).

SparkQA commented:
**[Test build #103338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103338/testReport)** for PR 23453 at commit [`d36c862`](https://github.com/apache/spark/commit/d36c86251946e88a36e2e977d8ae1a0ae815a8d5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103338/
Test PASSed.

squito commented:
merged to master, thanks @ankuriitg !
